crypto/cmp/cmp_msg.c

```diff
@@ -0,0 +1,997 @@
+    if (errorCode >= 0) {
+        msg->body->value.error->errorCode = ASN1_INTEGER_new();
+        if (!ASN1_INTEGER_set(msg->body->value.error->errorCode, errorCode))
+            goto err;
+    }
+    if (errorDetails != NULL)
+        msg->body->value.error->errorDetails =
+            sk_ASN1_UTF8STRING_deep_copy(errorDetails, ASN1_STRING_dup,
+                                         ASN1_STRING_free);
+
+    if (!unprotected && !ossl_cmp_msg_protect(ctx, msg))
+        goto err;
+    return msg;
+
+ err:
+    CMPerr(0, CMP_R_ERROR_CREATING_ERROR);
+    OSSL_CMP_MSG_free(msg);
+    return NULL;
+}
+
+/*
+ * OSSL_CMP_CERTSTATUS_set_certHash() calculates a hash of the certificate,
+ * using the same hash algorithm as is used to create and verify the
+ * certificate signature, and places the hash into the certHash field of a
+ * OSSL_CMP_CERTSTATUS structure. This is used in the certConf message,
+ * for example, to confirm that the certificate was received successfully.
+ */
+int ossl_cmp_certstatus_set_certHash(OSSL_CMP_CERTSTATUS *certStatus,
+                                     const X509 *cert)
+{
+    unsigned int len;
+    unsigned char hash[EVP_MAX_MD_SIZE];
+    int md_NID;
+    const EVP_MD *md = NULL;
+
+    if (!ossl_assert(certStatus != NULL && cert != NULL))
+        return 0;
+
+    /*-
+     * select hash algorithm, as stated in Appendix F. Compilable ASN.1 defs:
+     * the hash of the certificate, using the same hash algorithm
+     * as is used to create and verify the certificate signature
+     */
+    if (OBJ_find_sigid_algs(X509_get_signature_nid(cert), &md_NID, NULL)
+            && (md = EVP_get_digestbynid(md_NID)) != NULL) {
+        if (!X509_digest(cert, md, hash, &len))
+            goto err;
+        if (!ossl_cmp_asn1_octet_string_set1_bytes(&certStatus->certHash, hash,
+                                                   len))
+            goto err;
+    } else {
+        CMPerr(0, CMP_R_UNSUPPORTED_ALGORITHM);
+        return 0;
+    }
+
+    return 1;
+ err:
+    CMPerr(0, CMP_R_ERROR_SETTING_CERTHASH);
+    return 0;
+}
+
+/*
+ * TODO: handle potential 2nd certificate when signing and encrypting
+ * certificates have been requested/received
+ */
+OSSL_CMP_MSG *ossl_cmp_certConf_new(OSSL_CMP_CTX *ctx, int fail_info,
+                                    const char *text)
+{
+    OSSL_CMP_MSG *msg = NULL;
+    OSSL_CMP_CERTSTATUS *certStatus = NULL;
+    OSSL_CMP_PKISI *sinfo;
+
+    if (!ossl_assert(ctx != NULL && ctx->newCert != NULL))
+        return NULL;
+
+    if ((unsigned)fail_info > OSSL_CMP_PKIFAILUREINFO_MAX_BIT_PATTERN) {
+        CMPerr(0, CMP_R_FAIL_INFO_OUT_OF_RANGE);
+        return NULL;
+    }
+
+    if ((msg = ossl_cmp_msg_create(ctx, OSSL_CMP_PKIBODY_CERTCONF)) == NULL)
+        goto err;
+
+    if ((certStatus = OSSL_CMP_CERTSTATUS_new()) == NULL)
+        goto err;
+    if (!sk_OSSL_CMP_CERTSTATUS_push(msg->body->value.certConf, certStatus))
```

Comment:
We did the push to `msg` right away such that there is no need to bother with cleaning up `certStatus` in case of errors. I've added a respective comment.